demos/vc_issuer/src/main.rs

```diff
@@ -0,0 +1,521 @@
+            "Unsupported value for argument '{}': expected '{}', got '{}'",
+            expected_argument, expected_value, value
+        ));
+    }
+
+    let unexpected_arguments: Vec<&String> = arguments
+        .keys()
+        .into_iter()
+        .filter(|k| k.as_str() != expected_argument)
+        .collect();
+    if !unexpected_arguments.is_empty() {
+        return Err(format!(
+            "Unexpected arguments for credential {}: {:?}",
+            spec.credential_name, unexpected_arguments
+        ));
+    }
+    Ok(())
+}
+
+fn get_vc_consent_message(req: &Icrc21VcConsentMessageRequest) -> String {
+    format!(
+        "Issue credential '{}' with arguments:{}",
+        req.credential_spec.credential_name,
+        arguments_as_string(&req.credential_spec.arguments)
+    )
+}
+fn arguments_as_string(maybe_args: &Option<HashMap<String, ArgumentValue>>) -> String {
+    let mut arg_str = String::new();
+    let empty_args = HashMap::new();
+    let args = maybe_args.as_ref().unwrap_or(&empty_args);
+    for (key, value) in args {
+        arg_str.push_str(&format!("\n\t{}: {}", key, value));
+    }
+    arg_str
+}
+
+#[update]
+#[candid_method]
+fn add_employee(employee_id: Principal) -> String {
+    EMPLOYEES.with_borrow_mut(|employees| employees.insert(employee_id));
+    format!("Added employee {}", employee_id)
+}
+
+#[update]
+#[candid_method]
+fn add_graduate(graduate_id: Principal) -> String {
+    GRADUATES.with_borrow_mut(|graduates| graduates.insert(graduate_id));
+    format!("Added graduate {}", graduate_id)
+}
+
+fn main() {}
+
+fn add_signature(sigs: &mut SignatureMap, msg_hash: Hash, seed: Hash) {
+    let signature_expires_at = time().saturating_add(CERTIFICATE_VALIDITY_PERIOD_NS);
+    sigs.put(hash_bytes(seed), msg_hash, signature_expires_at);
+}
+
+fn get_signature(sigs: &SignatureMap, seed: Hash, msg_hash: Hash) -> Option<Vec<u8>> {
+    let certificate = data_certificate().unwrap_or_else(|| {
+        trap("data certificate is only available in query calls");
+    });
+    let witness = sigs.witness(hash_bytes(seed), msg_hash)?;
+
+    let witness_hash = witness.reconstruct();
+    let root_hash = sigs.root_hash();
+    if witness_hash != root_hash {
+        trap(&format!(
+            "internal error: signature map computed an invalid hash tree, witness hash is {}, root hash is {}",
+            hex::encode(witness_hash),
+            hex::encode(root_hash)
+        ));
+    }
+    let tree = ic_certified_map::labeled(LABEL_SIG, witness);
+    #[derive(Serialize)]
+    struct Sig<'a> {
+        certificate: ByteBuf,
+        tree: HashTree<'a>,
+    }
+
+    let sig = Sig {
+        certificate: ByteBuf::from(certificate),
+        tree,
+    };
+
+    let mut cbor = serde_cbor::ser::Serializer::new(Vec::new());
+    cbor.self_describe().unwrap();
+    sig.serialize(&mut cbor).unwrap();
+    Some(cbor.into_inner())
+}
+
+fn calculate_seed(principal: &Principal) -> Hash {
+    let dummy_salt = [5u8; 32];
```

Comment:
In a real implementation, should this be random data? If that's the case, and since this is meant as an example to copy, worth leaving a note?

Reply:
Correct, added a note.